Makefile

```diff
@@ -1,28 +1,40 @@
-BINDIR     := $(CURDIR)/bin
-DIST_DIRS  := find * -type d -exec
-TARGETS    := darwin/amd64 linux/amd64 linux/386 linux/arm linux/arm64 linux/ppc64le windows/amd64
-BINNAME    ?= helm
+PROJECTNAME = helm
 
-GOPATH     = $(shell go env GOPATH)
-DEP        = $(GOPATH)/bin/dep
-GOX        = $(GOPATH)/bin/gox
+# Required for globs to work correctly
+SHELL       = /bin/bash
+CHECK       = which
+NUL         = /dev/null
+BINNAME     = $(PROJECTNAME)
 
 # go option
-PKG        := ./...
-TAGS       :=
-TESTS      := .
-TESTFLAGS  :=
-LDFLAGS    := -w -s
-GOFLAGS    :=
-SRC        := $(shell find . -type f -name '*.go' -print)
-
-# Required for globs to work correctly
-SHELL      = /bin/bash
+PKG         := ./...
+TAGS        :=
+TESTS       := .
+TESTFLAGS   :=
+LDFLAGS     := -w -s
+GOFLAGS     :=
+SRC         = $(shell find . -type f -name '*.go' -print)
+
+GIT_COMMIT  = $(shell git rev-parse HEAD)
+GIT_SHA     = $(shell git rev-parse --short HEAD)
+GIT_TAG     = $(shell git describe --tags --abbrev=0 --exact-match 2>$(NUL))
+GIT_DIRTY   = $(shell test -n "`git status --porcelain`" && echo "dirty" || echo "clean")
+
+BINDIR      := $(CURDIR)/bin
+DIST_DIRS   := find * -type d -exec
+TARGETS     := darwin/amd64 linux/amd64 linux/386 linux/arm linux/arm64 linux/ppc64le windows/amd64
+
+ifeq ($(OS),Windows_NT)
+	BINNAME    = $(PROJECTNAME).exe
+	SHELL      = cmd.exe
```

Comment:
What is the `SHELL` environment variable used for? I see the `shell` make function is used to shell out. Am I missing something?

Reply:
From https://www.gnu.org/software/make/manual/html_node/Choosing-the-Shell.html:

> The program used as the shell is taken from the variable SHELL. If this variable is not set in your makefile, the program /bin/sh is used as the shell.

By setting this to cmd.exe (and other windows-equivalent variables), the Makefile will work without requiring cygwin or equivalent shell emulators installed.

Reply:
Prior to setting the `SHELL` to `cmd.exe` on Windows there a variables set that use the `shell` function. For example,

```
GIT_COMMIT  = $(shell git rev-parse HEAD)
```

Does `SHELL` being set to `cmd.exe` need to happen earlier in the script?